Update plugin to latest Gradle standards

jjohannes opened the pull request:
@REPLicated I was looking for a simple solution (before I build one myself) that scans a classpath for duplicates. Yours is exactly what I was looking for. I think something like this should be used by much more projects. 

I found a few things to update to make it follow latest Gradle standards and make it compatible with the new _Configuration Cache_ feature. Would you be interested in integrating this and release a new version?

**Background**
I am doing a series of videos about the Classpath and related concepts in Gradle (first part [here](https://www.youtube.com/watch?v=HqAp9JBl2_U&list=PLWQK2ZdV4Yl2k2OmC_gsjDpdIBTN0qqkE)) and need something like this there soon. And I maintain a [GitHub project](https://github.com/jjohannes/gradle-project-setup-howto) where I show how latest Gradle features may be used to get rid of dependency chaos - there I would also like to add this plugin (or something similar).


**About the changes in this PR:**
- Current Gradle version is now 7.5.1
- Minimal supported version is now 6.6
- The Task can now be cached (if Build Cache is active)
- The Task is now compatible with the Configuration Cache

Note: 6.6+ is needed tow write tasks as abstract classes with abstract getters.

The build now uses the Kotlin DSL so that the `embedded-kotlin` plugin can be used to compile/run against the Kotlin version embedded in Gradle (this is not supported in Groovy DSL right now).

jjohannes committed:
Update plugin to latest Gradle standards

- Current Gradle version is 7.5.1
- Minimal supported version is now 6.6

6.6+ is needed tow write tasks as abstract classes with abstract
getters.

The build now uses the Kotlin DSL so that the `embedded-kotlin` plugin
can be used to compile/run against the Kotlin version embedded in Gradle
(this is not supported in Groovy DSL right now).

REPLicated commented:
@jjohannes : Thank you for the contribution and the helpful description. I intend to test the changes in a project where I use this plugin and then release a new version in the next few days.

REPLicated commented on src/main/kotlin/io/fuchs/gradle/collisiondetector/CollisionDetectorPlugin.kt:
```diff
@@ -5,17 +5,18 @@ import org.gradle.api.Plugin
-            project.tasks.create(taskName, DetectCollisionsTask::class.java)
+        project.tasks.register(taskName, DetectCollisionsTask::class.java) {
+            // convention useful for most Jars, can be replaced completely using 'setExcludes()'
+            it.collisionFilter.exclude("META-INF/**", "module-info.class")
```
I have used this plugin the past in combination with the [Gradle Shadow](https://github.com/johnrengelman/shadow) plugin, which merges many JARs into a single one, as I have run into issues when Service Descriptor files under `META-INF/services` were colliding and overwriting each other.

There is an [option in that plugin ](https://imperceptiblethoughts.com/shadow/configuration/merging/#merging-service-descriptor-files) to enable merging of such files, which is helpful but does not happen by default.

This change would mean that we do not detect that situation (collision of these service descriptors) anymore by default. Can we easily include `META-INF/services/**` but exclude everything else in `META-INF`? In any case, I believe the convention here is useful.

jjohannes replied:
Good point. I did not think about such a use case.

> an we easily include META-INF/services/** but exclude everything else in META-INF?

This would be possible, but I think it would be an inconsistent default for the (standard?) use case where you check a "normal" classpath. Because then the duplicated service loader descriptors are allowed, expected, and used. Similar to how multiple `module-info.class` or `MANIFEST.MF` files are used.

In my opinion we should then either:

1. Use the default as suggested in this PR, which fits the use case of _"I apply the plugin and give it my runtime classpath and expect it **not** to fail if there are **no** problematic duplicates"_
2. Set no default at all, so that you always have to configure the plugin (depending on what you want to use it for)

I would still tend towards option 1, but we could also do 2.

jjohannes replied:
Another thing about conventions: 
I used the plugin a bit more, thinking about how it is most easy to explain what it does.

In that context, I got the idea that it would be a good default to set it to analyze the _runtimeClasspath_ when used in a Java project.

Then you can just apply the plugin, run `:detectCollisions` and immediately get results without the need to configure anything. I think this makes it very accessible.

But let me know if you don't like that, then I can remove it from the PR again.

REPLicated replied:
I agree with both your points on the conventions.

jjohannes committed:
Make sure Gradle uses Java Version from '.java-version' as target

This makes sure that even if another JDK version is used to build
the plugin (e.g. in an environment where jenv is not installed) the
version stored in the '.java-version' file is used as target.

jjohannes committed:
By default, analyze the 'runtime classpath' of java projects

jjohannes committed:
Add task group and description

REPLicated commented on build.gradle.kts:
```diff
@@ -0,0 +1,31 @@
+plugins {
+    `embedded-kotlin`
```
One last question: Is this `embedded-kotlin` documented anywhere? All I can find is this [entry](https://plugins.gradle.org/plugin/org.gradle.kotlin.embedded-kotlin/2.3.3) on the gradle plugin portal, which links to an archived github repository. I presume this repo contains the precursor to the modern kotlin DSL support.

I am familiar with .kts build scripts and also the `kotlin-dsl` plugin (documented [here](https://docs.gradle.org/current/userguide/kotlin_dsl.html#sec:kotlin-dsl_plugin)), but wondering about the plugin applied here.

(Digression: One aspect I find confusing when developing a Gradle plugin in Kotlin is the compatibility pitfalls regarding the kotlin stdlib version this plugin/gradle uses vs. the kotlin stdlib version used by a project that applies this plugin. If you ever get around to make a video about that, I would be interested).

jjohannes replied:
Yes the link there is outdated it seems. It all was integrated into https://github.com/gradle/gradle at some point.

> (Digression: One aspect I find confusing when developing a Gradle plugin in Kotlin is the compatibility pitfalls regarding the kotlin stdlib version this plugin/gradle uses vs. the kotlin stdlib version used by a project that applies this plugin. If you ever get around to make a video about that, I would be interested).

Yes I agree and it is the source of much confusion and discussion. Right now I would say it is best, when doing Gradle plugin development in Kotlin, to somehow use the version that is embedded in Gradle. Because that's the only version Gradle can use at runtime.

Looks like `embedded-kotlin` is not documented then. 😐 It is basically `kotlin.jvm` with the version embedded in Gradle. That's why I used it as it fits what you had before minus the version.

`kotlin-dsl` also applies `embedded-kotlin` under the hood and adds more conveniences on top. One thing it adds is a Kotlin compile plugin (SAM-with-receiver) that allows you to use `this` instead of `it` in closures. Which makes  the DSL more compact (and more similar to the Groovy DSL variant). We could also use that, but we might need to adjust the code in a few places then.

jjohannes commented:
Thank you @REPLicated for integrating this and releasing a new version!

jjohannes commented:
@REPLicated just want to let you know that I show the plugin in the video in case you ever want to use that: https://youtu.be/KocTqF0hO_8
